hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorizedraw/IsAuthorizedRawCall.java

```diff
@@ -0,0 +1,266 @@
+
+        // ECRECOVER's output always has status SUCCESS but the result Bytes are either empty or the recovered address.
+        if (ecrecoverResult.getOutput().isEmpty()) return false;
+        final var recoveredAddress =
+                ecrecoverResult.getOutput().slice(12).toArray(); // LAST 20 bytes are where the EVM address is
+
+        // ECRECOVER produced an address:  Must match our account's alias address
+        final var givenAddress = explicitFromHeadlong(address);
+        final var addressesMatch = 0 == Arrays.compare(recoveredAddress, givenAddress);
+
+        return addressesMatch;
+    }
+
+    /** Validate (native Hedera) ED25519 signature */
+    public boolean validateEdSignature(@NonNull final Account account, @NonNull final Key key) {
+        requireNonNull(account, "account");
+        requireNonNull(key, "key");
+
+        // Use of `com.swirlds.common.crypto.CryptographyHolder` straight from the Platform is deprecated:
+        // FUTURE: Get the `Cryptography` engine from the services app via the context (needs to be invented)
+
+        return CryptographyHolder.get()
+                .verifySync(messageHash, signature, key.ed25519OrThrow().toByteArray());
+    }
+
+    /** Encode the _output_ of our system contract: it's a boolean */
+    @NonNull
+    ByteBuffer encodedAuthorizationOutput(final boolean authorized) {
+        return IsAuthorizedRawTranslator.IS_AUTHORIZED_RAW.getOutputs().encodeElements(authorized);
+    }
+
+    /** Format our message hash and signature for input to the ECRECOVER precompile */
+    @NonNull
+    public Optional<byte[]> formatEcrecoverInput(@NonNull final byte[] messageHash, @NonNull final byte[] signature) {
+        requireNonNull(messageHash, "messageHash");
+        requireNonNull(signature, "signature");
+
+        // Signature:
+        //   [ 0;  31]  r
+        //   [32;  63]  s
+        //   [64;  64]  v (but possibly 0..1 instead of 27..28)
+
+        // From evm.codes, input to ECRECOVER:
+        //   [ 0;  31]  hash
+        //   [32;  63]  v == recovery identifier ∈ {27,28}
+        //   [64;  95]  r == x-value ∈ (0, secp256k1n);
+        //   [96; 127]  s ∈ (0; sep256k1n ÷ 2 + 1)
+
+        if (messageHash.length != 32 || signature.length != 65) return Optional.empty();
+        final var ov = reverseV(signature[64]);
+        if (ov.isEmpty()) return Optional.empty();
+
+        final byte[] result = new byte[128];
+        System.arraycopy(messageHash, 0, result, 0, 32); // hash
+        result[63] = ov.get(); //                           v
+        System.arraycopy(signature, 0, result, 64, 32); //  r
+        System.arraycopy(signature, 32, result, 96, 32); // s
+
+        return Optional.of(result);
+    }
+
+    /** Make sure v ∈ {27, 28} - but after EIP-155 it might come in with a chain id ... */
+    @NonNull
+    public Optional<Byte> reverseV(@NonNull final byte v) {
```

Comment:
```suggestion
    public Optional<Byte> reverseV(final byte v) {
```

Primitive types cannot be null

Reply:
Eek.  Should have seen the IntelliJ warning.  TY